kratos/containers/distributed_sparse_graph.h

```diff
@@ -0,0 +1,427 @@
+        auto it = std::upper_bound(mCpuBounds.begin(), mCpuBounds.end(), RowIndex);
+
+        KRATOS_DEBUG_ERROR_IF(it == mCpuBounds.end()) <<
+            "row RowIndex " << RowIndex <<
+            " is not owned by any processor " << std::endl;
+
+        IndexType owner_rank = (it-mCpuBounds.begin()-1);
+
+        KRATOS_DEBUG_ERROR_IF(owner_rank < 0) <<
+            "row RowIndex " << RowIndex <<
+            " is not owned by any processor " << std::endl;
+
+        return owner_rank;
+
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+    void Clear()
+    {
+        mLocalGraph.Clear();
+        mNonLocalGraphs.clear();
+    }
+
+    void AddEntry(const IndexType RowIndex, const IndexType ColIndex)
+    {
+        if(IsLocal(RowIndex)){
+            mLocalGraph.AddEntry(RowIndex, ColIndex);
+        }
+        else{
+            mNonLocalGraphs[OwnerRank(RowIndex)].AddEntry(RowIndex, ColIndex);
+        }
+    }
+
+    template<class TContainerType>
+    void AddEntries(const IndexType RowIndex, const TContainerType& rColIndices)
+    {
+        if(IsLocal(RowIndex)){
+            mLocalGraph.AddEntries(RowIndex, rColIndices);
+        }
+        else{
+            mNonLocalGraphs[OwnerRank(RowIndex)].AddEntries(RowIndex, rColIndices);
+        }
+    }
+
+    template<class TIteratorType>
+    void AddEntries(const IndexType RowIndex,
+                    const TIteratorType& rColBegin,
+                    const TIteratorType& rColEnd
+                    )
+    {
+        if(IsLocal(RowIndex)){
+            mLocalGraph.AddEntries(RowIndex, rColBegin, rColEnd);
+        }
+        else{
+            mNonLocalGraphs[OwnerRank(RowIndex)].AddEntries(RowIndex, rColBegin, rColEnd);
+        }
+    }
+
+    template<class TContainerType>
+    void AddEntries(const TContainerType& rIndices)
+    {
+        for(auto I : rIndices)
+        {
+            if(IsLocal(I)){
+                mLocalGraph.AddEntries(I, rIndices);
+            }
+            else{
+                mNonLocalGraphs[OwnerRank(I)].AddEntries(I, rIndices);;
+            }
+        }
+    }
+
+    void AddEntries(DistributedSparseGraph& rOtherGraph)
+    {
+        //TODO
+    }
+
+    void AddEntries(SparseGraph& rOtherGraph)
+    {
+        //TODO
+        for(const auto& item: rOtherGraph.GetGraph())
+        {
+            AddEntries(item.first, item.second);
+        }
+    }
+
+    void Finalize()
+    {
+        std::vector<int> send_list;
+        for(auto& item : mNonLocalGraphs)
+            send_list.push_back(item.first);
+
+        auto colors = MPIColoringUtilities::ComputeCommunicationScheduling(send_list, mrComm);
+
+        //sendrecv data
+        for(auto color : colors)
+        {
+            if(color >= 0) //-1 would imply no communication
+            {
+                //TODO: this can be made nonblocking
+                auto recv_graph = mrComm.SendRecv(mNonLocalGraphs[color], color, color);
```

Comment:
here i am using the "object" version of the SendRecv which is based on serializization. Maybe i should base this on basic types to ensure max performance